grouper/usecases/grant_permission_to_service_account.py

```diff
@@ -0,0 +1,155 @@
+        self, permission, argument, service, message
+    ):
+        # type: (str, str, str, str) -> None
+        pass
+
+    @abstractmethod
+    def grant_permission_to_service_account_failed_permission_not_found(self, permission, service):
+        # type: (str, str) -> None
+        pass
+
+    @abstractmethod
+    def grant_permission_to_service_account_failed_service_account_not_found(self, service):
+        # type: (str) -> None
+        pass
+
+    @abstractmethod
+    def granted_permission_to_service_account(self, permission, argument, service):
+        # type; (str, str, str) -> None
+        pass
+
+
+class GrantPermissionToServiceAccount(object):
+    def __init__(
+        self,
+        actor,  # type: str
+        ui,  # type: GrantPermissionToServiceAccountUI
+        permission_service,  # type: PermissionInterface
+        service_account_service,  # type: ServiceAccountInterface
+        user_service,  # type: UserInterface
+        group_service,  # type: GroupInterface
+        transaction_service,  # type: TransactionInterface
+    ):
+        # type: (...) -> None
+        self.actor = actor
+        self.ui = ui
+        self.permission_service = permission_service
+        self.service_account_service = service_account_service
+        self.user_service = user_service
+        self.group_service = group_service
+        self.transaction_service = transaction_service
+
+    def can_grant_permissions_for_service_account(self, service):
+        # type: (str) -> bool
+        if not self.service_account_service.service_account_is_enabled(service):
+            return False
+
+        # If the actor is a permission admin, they can grant any permission to the service account.
+        # Otherwise, they have to be a member of the owning group.
+        if self.service_account_service.service_account_exists(self.actor):
+            return self.service_account_service.service_account_is_permission_admin(self.actor)
+        elif self.user_service.user_is_permission_admin(self.actor):
+            return True
+        else:
+            owner = self.service_account_service.owner_of_service_account(service)
+            return owner in self.user_service.groups_of_user(self.actor)
+
+    def grant_permission_to_service_account(self, permission, argument, service):
+        # type: (str, str, str) -> None
+        if not self.service_account_service.service_account_is_enabled(service):
+            self.ui.grant_permission_to_service_account_failed_service_account_not_found(service)
+            return
+
+        valid, error = self.permission_service.is_valid_permission_argument(permission, argument)
+        if not valid:
+            assert error
+            self.ui.grant_permission_to_service_account_failed_invalid_argument(
+                permission, argument, service, error
+            )
+            return
+
+        # If the actor is a permission admin, they can grant any permission to the service account.
```

Comment:
Oof, yea, validation checks are something I've had a lot of trouble finding a good place for in hexagonal. Having this duplicated login with `can_grant_permissions_for_service_account` just so we can call the appropriate callbacks feels yucky, but I don't see a better structure given the conventions we're following :/

Reply:
I tried to find some way to use a helper function, but the deeper problem is that the logic is very slightly different because it has to take the argument into account in the actual change but not in the pre-check (since the argument is not known in the pre-check). But that's only relevant if the user is not a permission admin, since permission admins are allowed to grant permissions not held by the group.

I'm sure there's some better structure here that allows sharing more of the code, but I couldn't come up with it and decided to live with this mess. (I think it's likely that we're going to change this logic in the future anyway.)